### PR TITLE
a safe msgpack compressor

### PR DESCRIPTION
**Pull request by maxtaco**
- compresses only static map keys, never any values, so we don't have to worry about leaking sensitive tokens or IDs
- but will compress random values if they are on a whitelist
- works with some basic test cases
- needs some cleanups and more testing

### CODE REVIEW

**Comment by maxtaco** [comment]
Preliminary PR, for discussion purposs.

File: go/msgpackzip/buf [hunk 1]
```diff
@@ -0,0 +1,238 @@
+package safedecoding
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/buf]: unused i'll nuke it

File: go/msgpackzip/compress_test.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+func TestCompressWithWhitelist(t *testing.T) {
+	d := func(s string) []byte {
+		ret, _ := hex.DecodeString(s)
+		return ret
+	}
+
+	// This white list was devined using the above test, TestReportValuesFrequencies
+	wl := NewValueWhitelist()
+	wl.AddString("team1")
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress_test.go]: No, you just "know" from the context what the values are that need to be added.  Complicated, I agree.

File: go/msgpackzip/compress.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+	}
+	freqsSorted, err := c.sortFrequencies(freqs)
+	if err != nil {
+		return nil, err
+	}
+	return freqsSorted, nil
+}
+
+// run the compressor on the given input. Can be used only once
+// per instantiation of compressor object.
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress.go]: nah, it's kinda of an internal implementation detail that doesn't get exported.

File: go/msgpackzip/compress.go [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+	var i int
+	for k, v := range freqs {
+		ret[i] = Frequency{k, v}
+		i++
+	}
+	sort.SliceStable(ret, func(i, j int) bool { return ret[i].Freq > ret[j].Freq })
+	return ret, nil
+}
+
+// frequenciesToMao converts a sorted vectors of frequencies to a map (key -> uint),
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress.go]: :+1:

File: go/msgpackzip/compress.go [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+		return nil, err
+	}
+
+	return compressedKeymap, nil
+}
+
+// outputFinalProduct is the final pass output routine. It outputs the wrapper
+// 3-value array, the version prefix, the encoded data, and the compressed, encoded
+// keyMap.
+func (c *compressor) outputFinalProduct(version int, data []byte, compressedKeymap []byte) (output []byte, err error) {
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress.go]: 👍

File: go/msgpackzip/compress_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/stretchr/testify/require"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+)
+
+type testVector struct {
+	name   string
+	data   string
+	inFile bool
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress_test.go]: nah!

[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress_test.go]: :-1:

File: go/msgpackzip/compress.go [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+	var ret outputter
+	err = ret.outputArrayPrefix(msgpackIntFromUint(uint(3)))
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/msgpackzip/compress.go]: don't understand?

**Comment by joshblum** [comment]
close in favor of https://github.com/keybase/client/pull/14756?

**Comment by maxtaco** [comment]
See here: https://github.com/keybase/msgpackzip

**Pull request closed without merging**